DTS API close out for 2.3

mbolivar-nordic opened the pull request:
DTS changes for 2.3.

This PR is attempting to close out the remaining issues discussed in the last dev-review, namely being able to access data for *all* nodes regardless of their status.

This has knock-on effects in the API which cause various renames.

We're going to defer a few more of them until after code freeze to avoid breaking CI any more.

mbolivar-nordic committed:
doc: write documentation about DT-based device instantiation

Add more HOWTO information for the two current devicetree-based device
instantiation styles, and a bit more information on how to create
devices that depend on others.

Point to this from the Kconfig tips page, since it is meant as a
replacement for existing Kconfig practice.

Update macros.bnf.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

galak commented on include/devicetree.h:
```diff
@@ -1396,16 +1401,6 @@
- * @return 1 if the devicetree has that numbered instance,
- *         0 otherwise.
- */
-#define DT_HAS_DRV_INST(inst) DT_HAS_NODE_STATUS_OKAY(DT_DRV_INST(inst))
```
Shouldn't this become `DT_HAS_DRV_INST_STATUS_OKAY(inst)`

galak reviewed (commented):
Need decide if we want to phase out compat2enabled or tweak its semantics.  I think changing the semantics is ok since the only status values we've really dealt with are (no status, status = "okay", status = "ok", and status = "disabled".  I'm also good if we want to deprecate this and introduce something new.

mbolivar-nordic commented:
> Need decide if we want to phase out compat2enabled or tweak its semantics. I think changing the semantics is ok since the only status values we've really dealt with are (no status, status = "okay", status = "ok", and status = "disabled". I'm also good if we want to deprecate this and introduce something new.

We agreed in dev-review that whatever we do, we can't break gen_legacy_defines.py. It uses compat2enabled, so we can't change its semantics without also making changes to gen_legacy_defines.py, which seems a bit reckless this late in the release cycle.

I'm adding new functionality specifically to be able to leave the legacy scripts alone.

galak commented on drivers/adc/adc_shell.c:
```diff
@@ -476,13 +453,13 @@ SHELL_STATIC_SUBCMD_SET_CREATE(sub_adc_cmds,
 
 SHELL_STATIC_SUBCMD_SET_CREATE(sub_adc,
-#if DT_HAS_DRV_INST(0)
+#if DT_HAS_NODE_STATUS_OKAY(DT_DRV_INST(0))
```
I don't think this is correct.  if we have inst0 - ok, inst1 - disabled, inst2-ok, inst3-ok.  The DT_INST_FOREACH_OKAY will get inst0, inst2, inst3.  But dont here we aren't going to match.

mbolivar-nordic replied:
changed to:

```
#define ADC_SHELL_COMMAND(inst) \
	SHELL_CMD(ADC_##inst, &sub_adc_cmds, "ADC_" #inst, NULL),

SHELL_STATIC_SUBCMD_SET_CREATE(
	sub_adc,
	DT_INST_FOREACH_OKAY(ADC_SHELL_COMMAND)
	SHELL_SUBCMD_SET_END /* Array terminated. */
);
```

galak commented on drivers/clock_control/clock_control_mcux_pcc.c:
```diff
@@ -67,7 +67,12 @@ static const struct clock_control_driver_api mcux_pcc_api = {
 
-#if DT_HAS_DRV_INST(0)
+#define CHECK_INSTANCE_NUMBER(inst) \
+	BUILD_ASSERT((inst) == 0, "unsupported temp instance");
```
What's the thinking here?

mbolivar-nordic replied:
Just trying to move quickly. The driver is single-instance and was hard-coded to instance number 0, so if this blows up, the devicetree is violating the "there's just one instance" assumption and this driver needs to be looked at again.

carlescufi commented on boards/arm/frdm_k64f/pinmux.c:
```diff
@@ -123,7 +123,7 @@ static int frdm_k64f_pinmux_init(struct device *dev)
 
 #if DT_HAS_NODE_STATUS_OKAY(DT_NODELABEL(ftm3)) && \
-    DT_NODE_HAS_COMPAT(DT_NODELABEL(ftm3), nxp_kinetis_ftm_pwm)
+    DT_NODE_HAS_COMPAT_STATUS_OKAY(DT_NODELABEL(ftm3), nxp_kinetis_ftm_pwm)
```
Wouldn't it be better to have the `OKAY` as a parameter to the macro? The name is pretty ugly and to top that it's inconsistent with `DT_INST_FOREACH_OKAY`

mbolivar-nordic replied:
Maybe it's better to make DT_INST_FOREACH_OKAY DT_INST_FOREACH_STATUS_OKAY instead. The STATUS_OKAY thing is being used elsewhere.

galak commented:
> > Need decide if we want to phase out compat2enabled or tweak its semantics. I think changing the semantics is ok since the only status values we've really dealt with are (no status, status = "okay", status = "ok", and status = "disabled". I'm also good if we want to deprecate this and introduce something new.
> 
> We agreed in dev-review that whatever we do, we can't break gen_legacy_defines.py. It uses compat2enabled, so we can't change its semantics without also making changes to gen_legacy_defines.py, which seems a bit reckless this late in the release cycle.
> 
> I'm adding new functionality specifically to be able to leave the legacy scripts alone.

We should probably add a comment in compat2enabled and node.enabled about these only existing for gen_legacy_defines.py.  Do we want to call this compat2avail to match the linux API?

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -359,6 +353,15 @@ def write_compatibles(node):
 
 
+def write_status(node):
+    status = node.props.get('status')
```
Use node.status now.

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -359,6 +353,15 @@ def write_compatibles(node):
+    status = node.props.get('status')
+
+    if status is None:
+        return
```
We can either remove this check or make it a warning as `node.status` should never be none.

mbolivar-nordic committed:
dts: clean up some redundant DT checks

These are redundantly checking a node's status twice.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mbolivar-nordic committed:
devicetree: allow access to all nodes

Usually, we want to operate only on "available" device
nodes ("available" means "status is okay and a matching binding is
found"), but that's not true in all cases.

Sometimes we want to operate on special nodes without matching
bindings, such as those describing memory.

To handle the distinction, change various additional devicetree APIs
making it clear that they operate only on available device nodes,
adjusting gen_defines and devicetree.h implementation details
accordingly:

- emit macros for all existing nodes in gen_defines.py, regardless
  of status or matching binding
- rename DT_NUM_INST to DT_NUM_INST_STATUS_OKAY
- rename DT_NODE_HAS_COMPAT to DT_NODE_HAS_COMPAT_STATUS_OKAY
- rename DT_INST_FOREACH to DT_INST_FOREACH_STATUS_OKAY
- rename DT_ANY_INST_ON_BUS to DT_ANY_INST_ON_BUS_STATUS_OKAY
- rewrite DT_HAS_NODE_STATUS_OKAY in terms of a new DT_NODE_HAS_STATUS
- resurrect DT_HAS_NODE in the form of DT_NODE_EXISTS
- remove DT_COMPAT_ON_BUS as a public API
- use the new default_prop_types edtlib parameter

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
> Do we want to call this compat2avail to match the linux API?

Do you also want to change all the STATUS_OKAY macros to say AVAIL? Would be best to keep those consistent in my opinion.

pabigot commented:
> > Do we want to call this compat2avail to match the linux API?
> 
> Do you also want to change all the STATUS_OKAY macros to say AVAIL? Would be best to keep those consistent in my opinion.

I think I support that.  Or `_AVAILABLE`.

From what I can see Linux really uses this to mean `status = "okay"`, so we should keep that interpretation.  To handle #19448 we'll need some other technical term for things for which devices should be generated, but that should not be started in the boot-time driver initialization.  I don't think there's a term for that in Linux because everything in the devicetree blob is "available" (in the general sense).

mbolivar-nordic commented:
I'm going to leave this as STATUS_OKAY to avoid yet more churn unless someone objects.

galak commented on scripts/dts/edtlib.py:
```diff
@@ -142,7 +142,8 @@ class EDT:
     def __init__(self, dts, bindings_dirs, warn_file=None,
-                 warn_reg_unit_address_mismatch=True):
+                 warn_reg_unit_address_mismatch=True,
+                 default_prop_types=None):
```
This means ever user (kconfiglib, sanitycheck, etc) has to define the same set of default props?  That doesn't make sense to me.

galak reviewed (commented):
Would be good to update doc/guides/dts/macros.bnf in this PR so we don't forget.

galak committed:
tests: devicetree: Add explicit has_status checks

Add some tests for DT_NODE_HAS_STATUS

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mbolivar-nordic commented:
> Would be good to update doc/guides/dts/macros.bnf in this PR so we don't forget.

Pushed various docs updates.

galak commented on scripts/dts/edtlib.py:
```diff
@@ -160,12 +161,17 @@ def __init__(self, dts, bindings_dirs, warn_file=None,
           node
+
+        default_prop_types (default: False):
+          If True, default property types will be used when a node has no
```
Can we default this to True and update the old script to call it with False.

mbolivar-nordic replied:
> Can we default this to True and update the old script to call it with False.

done

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -36,7 +36,8 @@ def main():
                          warn_reg_unit_address_mismatch=
-                             "-Wno-simple_bus_reg" not in args.dtc_flags)
+                             "-Wno-simple_bus_reg" not in args.dtc_flags,
+                         default_prop_types=False)
```
Don't you want to remove default_prop_types here.

mbolivar-nordic replied:
It becomes True later on in the series (`devicetree: allow access to all nodes`). I'm just trying to preserve bisectability of behavior.

galak commented on scripts/dts/edtlib.py:
```diff
@@ -160,12 +161,17 @@ def __init__(self, dts, bindings_dirs, warn_file=None,
           the address of the first entry does not match the unit address of the
           node
+
+        default_prop_types (default: False):
```
Need to fix comment to match the code `(default: False) -> (default: True)`

mbolivar-nordic replied:
Thanks for catching that

mbolivar-nordic commented:
Will fix the shippable issues in the morning.

galak committed:
drivers: move nxp_kinetis_temp to for-each style

Move to FOREACH to prepare for an upcoming patch.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
We need to look at DT_ANY_INST_ON_BUS_STATUS_OKAY  -- drivers/sensor/iis2dlpc fails to build on frdm_k64f.  Since DT_ANY_INST_ON_BUS_STATUS_OKAY(spi) is true, but INST(0) is not enabled.

galak commented:
We need fix 25059 and I've got a fix for the nxp_kinetis_temp

galak commented:
> We need to look at DT_ANY_INST_ON_BUS_STATUS_OKAY -- drivers/sensor/iis2dlpc fails to build on frdm_k64f. Since DT_ANY_INST_ON_BUS_STATUS_OKAY(spi) is true, but INST(0) is not enabled.

So I think for singleton drivers we need something that gives back the list of OKAY INST numbers and than we can use the first item in that list for the singletons

erwango commented on drivers/can/can_stm32.c:
```diff
@@ -20,10 +20,8 @@ LOG_MODULE_DECLARE(can_driver, CONFIG_CAN_LOG_LEVEL);
-	DT_NODE_HAS_COMPAT(DT_NODELABEL(can1), st_stm32_can) && \
-	DT_HAS_NODE_STATUS_OKAY(DT_NODELABEL(can2)) && \
-	DT_NODE_HAS_COMPAT(DT_NODELABEL(can2), st_stm32_can)
+#if DT_NODE_HAS_COMPAT_STATUS_OKAY(DT_NODELABEL(can1), st_stm32_can) && \
```
While it should work, it doesn't:
`sanitycheck -b -T  samples/drivers/can/ -p olimex_stm32_p405`
```
/local/mcu/zephyrproject/zephyr/drivers/can/can_stm32.c:25:2: error: #error Simultaneous use of CAN_1 and CAN_2 not supported yet
   25 | #error Simultaneous use of CAN_1 and CAN_2 not supported yet
      |  ^~~~~

```

mbolivar-nordic replied:
@erwango fixed now, please take another look.

galak committed:
devicetree: DT_NODE_HAS_COMPAT_STATUS_OKAY -> DT_NODE_HAS_COMPAT_STATUS

Swap this out and make the status a parameter.
Leave a couple of cases of DT_NODE_HAS_COMPAT().

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>
Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
gen_defines: play tricks to benefit DT_INST users

Even though it is about to be done for sound technical reasons, a
subsequent patch adding access to all device nodes at the last minute
in the 2.3 release is going to be playing a bit of a fast one on
the Zephyr community, especially users of DT_INST APIs.

In particular, instance numbers are currently allocated only to
enabled nodes, but that will not be true soon: *every* node of a
compatible will be allocated an instance number, even disabled ones.

This is especially unfortunate for drivers and applications that
expect singletons of their compatibles, and use DT_INST(0, ...) to
mean "the one enabled instance of my compatible".

To avoid gratuitous breakage, let's prepare for that by sorting each
edt.compat2nodes sub-list so that enabled instances always come before
disabled ones.

This doesn't break any API guarantees, because there basically *are*
no ordering guarantees, in part precisely to give us the flexibility
to do things like this. And it does help patterns that use instances 0
through N-1, including the important singleton case.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

pabigot reviewed (approved):
Haven't looked closely at the documentation yet but figure that can be tuned if necessary.

mbolivar-nordic commented:
> Haven't looked closely at the documentation yet but figure that can be tuned if necessary.

Yep, I promise to go over the documentation with a fine tooth comb before the release and correct any errors I find during the stabilization/documentation period. Thanks.

ioannisg commented on doc/guides/dts/howtos.rst:
```diff
@@ -104,8 +104,7 @@ a node identifier and pass it to ``DT_LABEL`` to get the right string to pass
 If you're having trouble, see :ref:`dt-trouble`. The first thing to check is
-that the node is enabled (``status = "okay"``) and has a matching binding, like
-this:
+that the node is has ``status = "okay"`` and has a matching binding, like this:
```
remove "is"

ioannisg commented on doc/guides/dts/howtos.rst:
```diff
@@ -114,11 +113,11 @@ this:
 
-If you see the ``#error`` output, something is wrong with either your
-devicetree or bindings.
+If you see the ``#error`` output, either something is wrong with either your
```
remove second either in the line

pabigot replied:
No, this is correct.  There are two optional subordinate clauses, one of which requires its own `either`.

pabigot replied:
Or rather, it's correct with two eithers, just the first, just the second, or neither.  IMO.  English is pretty flexible.

ioannisg reviewed (approved):
Looks ok to me
Pls, address some minor doc-related comments.
